src/test/java/UploadReceiptServletTest.java

```diff
@@ -0,0 +1,331 @@
+  private static final double PRICE = 5.89;
+  private static final String STORE = "McDonald's";
+  private static final AnalysisResults ANALYSIS_RESULTS = new AnalysisResults(RAW_TEXT.getValue());
+
+  private static final String IMAGE_URL = "/serve-image?blob-key=" + BLOB_KEY.getKeyString();
+  private static final String LIVE_SERVER_BASE_URL =
+      "https://capstone-receipt-step-2020.uc.r.appspot.com:80";
+  private static final String LIVE_SERVER_SCHEME = "https";
+  private static final String LIVE_SERVER_NAME = "capstone-receipt-step-2020.uc.r.appspot.com";
+  private static final int LIVE_SERVER_PORT = 80;
+  private static final String LIVE_SERVER_CONTEXT_PATH = "";
+  private static final String DEV_SERVER_SCHEME = "http";
+  private static final String DEV_SERVER_NAME = "0.0.0.0";
+  private static final int DEV_SERVER_PORT = 80;
+  private static final String DEV_SERVER_CONTEXT_PATH = "";
+
+  @Before
+  public void setUp() {
+    MockitoAnnotations.initMocks(this);
+    helper.setUp();
+    datastore = DatastoreServiceFactory.getDatastoreService();
+
+    // Create a fixed time clock that always returns the same instant.
+    clock = Clock.fixed(Instant.parse(INSTANT), ZoneId.systemDefault());
+
+    servlet = new UploadReceiptServlet(logger, blobstoreService, blobInfoFactory, datastore, clock);
+  }
+
+  @After
+  public void tearDown() {
+    helper.tearDown();
+  }
+
+  @Test
+  public void doGetReturnsBlobstoreUploadUrl() throws IOException {
+    StringWriter stringWriter = new StringWriter();
+    PrintWriter writer = new PrintWriter(stringWriter);
+    when(response.getWriter()).thenReturn(writer);
+
+    UploadOptions uploadOptions =
+        UploadOptions.Builder.withMaxUploadSizeBytesPerBlob(MAX_UPLOAD_SIZE_BYTES);
+    when(blobstoreService.createUploadUrl("/upload-receipt", uploadOptions)).thenReturn(UPLOAD_URL);
+
+    servlet.doGet(request, response);
+    writer.flush();
+    Assert.assertTrue(stringWriter.toString().contains(UPLOAD_URL));
+  }
+
+  @Test
+  public void doPostUploadsReceiptToDatastoreLiveServer() throws IOException {
+    // Add mock blob to Blobstore.
+    Map<String, List<BlobKey>> blobs = new HashMap<>();
+    blobs.put("receipt-image", Arrays.asList(BLOB_KEY));
+    when(blobstoreService.getUploads(request)).thenReturn(blobs);
+    BlobInfo blobInfo = new BlobInfo(
+        BLOB_KEY, "image/jpeg", new Date(), VALID_FILENAME, IMAGE_SIZE_1MB, HASH, null);
+    when(blobInfoFactory.loadBlobInfo(BLOB_KEY)).thenReturn(blobInfo);
+
+    when(request.getParameter("label")).thenReturn(LABEL);
+
+    // Stub request with URL components.
+    when(request.getScheme()).thenReturn(LIVE_SERVER_SCHEME);
+    when(request.getServerName()).thenReturn(LIVE_SERVER_NAME);
+    when(request.getServerPort()).thenReturn(LIVE_SERVER_PORT);
+    when(request.getContextPath()).thenReturn(LIVE_SERVER_CONTEXT_PATH);
+    String absoluteUrl = LIVE_SERVER_BASE_URL + IMAGE_URL;
```

Comment:
I think it would be better to have an ABSOLUTE_URL constant constructed from the other two constants rather than constructing that string in two different places.